tsdb/head_append.go

```diff
@@ -389,6 +383,67 @@ func (a *headAppender) Append(ref storage.SeriesRef, lset labels.Labels, t int64
 	return storage.SeriesRef(s.ref), nil
 }
 
+// AppendCreatedTimestamp appends a sample with 0 as its value.
+// AppendCreatedTimestamp does not support out-of-order appends.
+func (a *headAppender) AppendCreatedTimestamp(ref storage.SeriesRef, lset labels.Labels, t int64) (storage.SeriesRef, error) {
+	s := a.head.series.getByID(chunks.HeadSeriesRef(ref))
+	if s == nil {
+		var err error
+		s, err = a.getOrCreate(lset)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	s.Lock()
+	isOOO, _, err := s.appendable(t, 0, a.headMaxt, a.minValidTime, a.oooTimeWindow)
+	if err == nil {
+		s.pendingCommit = true
+	}
+	s.Unlock()
+	if err != nil {
+		return 0, err
+	}
+
+	if !isOOO {
+		if t > a.maxt {
```

Comment:
Prefer more complex cases further (e.g. filter out OOO case first) - for readability